packages/SwingSet/src/controller/upgradeSwingset.js

```diff
@@ -204,6 +207,154 @@ export const upgradeSwingset = kernelStorage => {
+          notifies.set(kpid, []);
+        }
+        notifies.get(kpid).push(vatID);
+      }
+    }
+    const [accHead, accTail] = JSON.parse(getRequired('acceptanceQueue'));
+    for (let p = accHead; p < accTail; p += 1) {
+      const rq = JSON.parse(getRequired(`acceptanceQueue.${p}`));
+      if (rq.type === 'notify') {
+        const { vatID, kpid } = rq;
+        assert(vatID);
+        assert(kpid);
+        if (!notifies.has(kpid)) {
+          notifies.set(kpid, []);
+        }
+        notifies.get(kpid).push(vatID);
+      }
+    }
+    console.log(` - pending notifies:`, notifies);
+
+    // cache of known-settled kpids: will grow to num(kpids)
+    const settledKPIDs = new Set();
+    const nonSettledKPIDs = new Set();
+    const isSettled = kpid => {
+      if (settledKPIDs.has(kpid)) {
+        return true;
+      }
+      if (nonSettledKPIDs.has(kpid)) {
+        return false;
+      }
+      const state = kvStore.get(`${kpid}.state`);
+      // missing state means the kpid is deleted somehow, shouldn't happen
+      assert(state, `${kpid}.state is missing`);
+      if (state === 'unresolved') {
+        nonSettledKPIDs.add(kpid);
+        return false;
+      }
+      settledKPIDs.add(kpid);
+      return true;
+    };
+
+    // walk vNN.c.kpNN for all vats, for each one check the
+    // kpNN.state, for the settled ones check for a pending notify,
+    // record the ones without a pending notify
+
+    const buggyKPIDs = []; // [kpid, vatID]
+    for (const vatID of allVatIDs) {
+      const prefix = `${vatID}.c.`;
+      const len = prefix.length;
+      const ckpPrefix = `${vatID}.c.kp`;
+      for (const key of enumeratePrefixedKeys(kvStore, ckpPrefix)) {
+        const kpid = key.slice(len);
+        if (isSettled(kpid)) {
+          const n = notifies.get(kpid);
+          if (!n || !n.includes(vatID)) {
+            // there is no pending notify
+            buggyKPIDs.push([kpid, vatID]);
+          }
+        }
+      }
+    }
+    console.log(` - found ${buggyKPIDs.length} buggy kpids, enqueueing fixes`);
+
+    // now fix it. The bug means we failed to delete the c-list entry
+    // and decref it back when the promise was rejected. That decref
+    // would have pushed the kpid onto maybeFreeKrefs, which would
+    // have triggered a refcount check at end-of-crank, which might
+    // have deleted the promise records (if nothing else was
+    // referencing the promise, like arguments in messages enqueued to
+    // unresolved promises, or something transient on the
+    // run-queue). Deleting those promise records might have decreffed
+    // krefs in the rejection data (although in general 9039 rejects
+    // those promises with non-slot-bearing DisconnectionObjects).
+    //
+    // To avoid duplicating a lot of kernel code inside this upgrade
+    // handler, we do the simplest possible thing: enqueue a notify to
+    // the upgraded vat for all these leftover promises. The new vat
```

Comment:
I am not the biggest fan of the kernel adding things to the run queue arbitrarily like this.

At upgrade time we may actually be in the middle of a run, and the reactions to these notifies will become part of that run. I would much prefer if this became a separate "upgrade run". This case is actually a good candidate for it. We can run the remediation anytime (as long as the targeted vats have not terminated), as the list of kpid is not going to grow anymore.

Maybe `upgradeSwingset` can return a closure that performs either the walk + enqueuing, or just the enqueuing from a walk that was already done. Or an alternative would be to return a description of what should be done (the list of queue items to add), and require the host to call a controller function that understands what to do. We already have a `queueTo` functions on the controller, so it's not a stretch to add a `controller.queueUpgradeAction(actionReturnedByUpgrade)`.

In either case, the host would perform this action when the previous run has completed, and then perform a new run just for this. It doesn't even have to be a "must run this block" with an unlimited policy, although that may still be preferable.

I think this would allow us to limit the amount of boundary piercing we have to do (no need to get stats here, or any functions that mutate the state kept by the keeper). It also sets up a framework for injecting actions at upgrade time, which have possibly not been self discovered but instead required offline processing (collection of some virtual data cycles comes to mind)

Reply:
@mhofman and I walked through the code. The benefits of his scheme is that the post-chain-halting-upgrade execution slogs will show three distinct phases: catch-up, then swingset upgrade actions (e.g. injected `dispatch.notify` from the promise remediations), then finally core-eval upgrade proposals (like installing bundles and upgrading vats). Without this scheme, the order would be: initial catch-up cranks (anything that was already in the run-queue when the chain halted), swingset-upgrade/remediation cranks, then later catch-up cranks (anything triggered by the initial catch-up cranks), then core-eval upgrade proposals.

The drawbacks of this scheme are the complexity of additional controller APIs, an extra operational burden on host-app authors (swingset no longer upgrades itself by default, it needs help from the caller), and changes to cosmic-swingset to drive it (which is our least-well tested package, so I try to minimize changes to it).

I agreed to pursue his approach. The details:

On the kernel side:

* `upgradeSwingset()` returns an "upgrade actions list", which is opaque (to the host app) and basically contains a list of run-queue event records
  * upgradeSwingset must increment the refcounts for any krefs in those records, just as it would if it pushed them onto the run-queue directly
* `controller.addEvents(upgradeActionsList)` will take that list and add it to the run-queue. It nominally increfs (as the items are pushed onto the run-queue) and decrefs (as the `upgradeActionsList` item is deleted), but it's a wash, so we won't actually modify refcounts in the function, we'll just have a comment explaining why we don't have to.
* The host app is now responsible for conveying the actions from `upgradeSwingset` to the controller. These actions must be injected during the first run after upgrade (i.e. the host app must not `hostStorage.commit()` before it calls `c.addEvents`). If they fail to do so, the kernel state will be corrupted. (I lament this additional burden)

On the cosmic-swingset side:

* we'll introduce a new `ActionType.INJECT_UPGRADE_ACTIONS`
* inside launch-chain.js `buildSwingset()`, the call to `upgradeSwingset` will grab the returned upgrade actions list and return it next to `coreProposals` and `controller`
* in `launch() > doBlockingSend() -> case ActionType.AG_COSMOS_INIT`, when we build `coreProposals`, we'll also look at the upgrade actions list, and if it is not empty, we'll build an `INJECT_UPGRADE_ACTIONS` around it and push that onto `runThisBlock`, before we call `doCoreProposals`
  * this induces a dependency: if SwingSet is upgraded (such that `upgradeSwingSet()` returns actions that need to be injected), but the golang side doesn't realize that we're performing a chain-halting upgrade (such that it fails to send an `AG_COSMOS_INIT`), then we'll get corrupted kernel state
  * it would be nice to catch this error case with an assertion, rather than allowing the kernel upgrade actions to be silently dropped, but I think that requires reacting to the *lack* of an `AG_COSMOS_INIT` event, which sounds difficult
  
Now, on normal reboots, `upgradeSwingset()` returns nothing (maybe `undefined`, maybe `[]`, we'd need to choose), and no `AG_COSMOS_INIT` is pushed, so the `runKernel()` code does it's normal budget-limited catchup and action-queue servicing.

But after a chain-halting upgrade, `upgradeSwingset()` might return a non-empty list (if the chain software upgrade included a kernel upgrade, and if that kernel upgrade included a state upgrade, and if that state upgrade produced events that needed injection). *And*, the golang side will emit an `AG_COSMOS_INIT` event at the beginning of the first post-upgrade block. If so, the `AG_COSMOS_INIT` handler will notice the non-empty swingset upgrade action list, and an `INJECT_UPGRADE_ACTIONS` event onto `runThisBlock` before it pushes anything else (like core-proposals) associated with the upgrade.

Then, when we reach `END_BLOCK`, the catch-up run will use `neverStop` (because `runThisBlock` is empty), and then it will service `runThisBlock`, which will start with the swingset injections. These will run to completion before processing any of the core-proposals, because they appear in a separate action than the `ActionType.CORE_EVAL` events that might follow.

The empty/non-empty state of `runThisBlock` (for each block) is a part of consensus, because it affects the runPolicy budget for the catchup run. So it is important that non-upgrade reboots do not spuriously populate it. This will be doubly ensured:
* only the `AG_COSMOS_INIT` handler can push onto `runThisBlock`
* normal reboots should not be changing kernel code, so even though `upgradeSwingset` is called on every reboot, normal reboots should not cause it to actually do an upgrade, so those calls will never return any actions to be injected.

Reply:
> * it fails to send an `AG_COSMOS_INIT`), then we'll get corrupted kernel state

There is always an init event (it happens for every restart). The event has flags indicating whether it's a consensus restart (bootstrap & upgrade info).

Like for core proposals, you can assert that swingset upgrade actions are only allowed on consensus restarts.

Reply:
I very much agree with the original comment, and support the proposed plan.

Reply:
...and further considering that we must not lose these remediation events, _even if there is enough leftover work to fill more than one block_, I prefer a strategy in which their presence is communicated to the controller so that it can insist that the very next `run()` sees them first and gets all the way through (i.e., hosts are free to control exactly how they are injected but are not free to ask for other work first—if they care about not interrupting leftover, then they must delay the upgrade until there isn't any).

This does raise the question of what to do with followup actions from the initial events, for which my preference would be to treat as normal, effectively being born as leftover work (while reserving the option for a future change that instead treats those as part of the queue-jumping remediation).

Reply:
In this case, the swing-store upgrade can be performed (gather the list of promises to notify rejections), then the controller built and the left over work executed, then the injected events from the upgrade can be run without comingling. This is possible in this case because the notify events remain valid (the state of the promise doesn't change further). Since we already do not remove events from the run queue when the vat terminates, even that edge case is handled by [simply dropping the notify if that happens](https://github.com/Agoric/agoric-sdk/blob/3007c001ee490c847987bc61d22acf9689611340/packages/SwingSet/src/kernel/kernel.js#L591-L594).

I think all this remains true in the general case of injected events. As long as injected events have a valid refcount keeping their slots alive, the deliveries they represent are akin to being held in a separate queue. I expect a similar pattern could be used to inject retire exports to trigger collection of cycles.

Reply:
Ok https://github.com/Agoric/agoric-sdk/pull/10337 is a secondary PR to this branch to address this. `upgradeSwingset()` is changed to record an internal list of the messages (notifies) that need to be injected, the new `controller.injectQueuedUpgradeEvents()` will do that, without revealing the messages to the caller. The host-app is newly responsible for calling that function after `upgradeSwingset()` (at least if it returns `modified === true`), and ideally before the next `commit()` (to avoid risking forgetting to do it), and before a `run()` (to get the work done promptly instead of leaving it for a future block). The cosmic-swingset code is changed to call it after a full cleanup/leftover run, but before any coreProposals, by adding a new `ActionType.KERNEL_UPGRADE_EVENTS` to the `runThisBlock` queue, but only when cosmos tells us that we're resuming from a chain-halting software upgrade.